src/main/java/io/dexi/config/Config.java

```diff
@@ -0,0 +1,187 @@
+package io.dexi.config;
+
+import org.apache.commons.configuration2.Configuration;
+import org.apache.commons.configuration2.FileBasedConfiguration;
+import org.apache.commons.configuration2.YAMLConfiguration;
+import org.apache.commons.configuration2.builder.FileBasedConfigurationBuilder;
+import org.apache.commons.configuration2.builder.ReloadingFileBasedConfigurationBuilder;
+import org.apache.commons.configuration2.builder.fluent.Parameters;
+import org.apache.commons.configuration2.builder.fluent.PropertiesBuilderParameters;
+import org.apache.commons.configuration2.ex.ConfigurationException;
+import org.apache.commons.lang.StringUtils;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ *  This class supports reading a "hierarchical" configuration file with sections like:
+ *
+ *  <b>YAML example</b>
+ *  <pre>
+ *  {@code
+ *     dexi:
+ *       baseUrl: http://localhost:3000/api/
+ *       apiKey: super-secret-key
+ *  }
+ *  </pre>
+ *
+ *  <b>JSON example</b>
+ *  <pre>
+ *  {@code
+ *      {
+ *          "dexi": {
+ *              "baseUrl": "http://localhost:3000/api/",
+ *              "apiKey": "super-secret-key"
+ *          }
+ *      }
+ *  }
+ *  </pre>
+ *
+ *  The file can be read from a local disk or from a URL. The class also supports reading environment variables whose
+ *  names start with "DEXI_APP_".
+ *
+ *  Configuration is read in the following order:
+ *  <ol>
+ *      <li>A local configuration file as specified by the {@code localConfigFile} parameter. The default is
+ *          {@code ~/.dexi/my-app.yml}.
+ *          <ul>
+ *              <li>Supported file formats are YAML (.yml), JSON (.json), XML (.xml) and INI (.ini).</li>
+ *          </ul>
+ *      </li>
+ *      <li>If an environment variable or system property named {@code DEXI_APP_CONFIG_URL_YML} is set, read a YAML (.yml)
+ *          file from the specified URL.</li>
+ *      <li>Read any {@code DEXI_APP_} environment variable. The format is: {@code DEXI_APP_&lt;section>_&lt;key> = &lt;value>}.</li>
+ *  </ol>
+ *
+ *  Values for duplicate keys within sections are overwritten by later keys.
+ *
+ */
+public class Config {
+
+    public static final String DEXI_APP_CONFIG_URL = "DEXI_APP_CONFIG_URL";
+
+    private static final String DEXI_APP_ENVIRONMENT_VARIABLE_PREFIX = "DEXI_APP_";
+
+    private static String localConfigFile = System.getProperty("user.home") + "/.dexi/my-app.yml";
```

Comment:
Default value - should probably be ".dexi/configuration.yml"

Reply:
Also this should probably be named something like "defaultLocalConfigFile"

Reply:
I actually prefer "my-app".

Will implement the "defaultLocalConfigFile" change.

Reply:
but my-app sounds a lot like something you should change ? E.g. "Hallo world"-like

Reply:
I don't think so, no. The point of the current implementation is that it allows the user to specify *any* name for the configuration file. The documentation then states what filename the class will look for, if a name is not specified.